Barclamp guide

jgrassler opened the pull request:
This pull requests adds a step-to-step guide for writing a new barclamp.

matrixik commented on doc/barclamp.md:
```diff
@@ -220,21 +221,21 @@ this phase development will have to move to your Crowbar admin node, though.
 
-* `chef/cookbooks/mybarclamp/metadata.rb`
-* `chef/cookbooks/mybarclamp/README.md`
+* `chef/cookbooks//metadata.rb`
```
Looks like here and in many other places you accidentally removed `mybarclamp` in this commit.

toabctl commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+# Barclamp development guide
+
+This document is an introduction to Barclamp development. It will guide you
```
A short description what a ```barclamp``` is would be nice in the intro

jgrassler replied:
Done.

toabctl commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+We will make heavy use of the `Cpatch`, `Fpatch` and `sync_crowbar` commands
+thus defined in the following sections. A short explanation of the three:
+
+* `Fpatch` will apply your whole fork against `/opt/dell` using `patch(1)`. You
```
Why not using ```barclamp_install``` with a given directory?

jgrassler replied:
Ah. That would do the trick, too, and be a lot simpler to boot. Are you currently using this approach? If so, are there any snags to watch out for?

jgrassler replied:
Hmm. Now that I think about it again, it's got the same problem as the copy-stuff-to-`/opt/dell` approach: if one doesn't rebase religiously and frequently, old code in _other_ barclamps that is no longer around in `master` might creep in, leading to all sorts of weird problems that have nothing to do with the barclamp being developed.

For completeness' sake I mentioned the `barclamp_install` approach alongside the `cp` one now.

toabctl commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+aspects of barclamp implementation in the order you are likely to encounter
+them. They are meant to be read in parallel to developing your own barclamp.
+
+## Development Phases
```
Do you know that there is a (broken) script to create a new barclamp? See https://github.com/crowbar/crowbar/blob/master/scripts/barclamp_create.rb .

jgrassler replied:
No, that one is news to me :-)

I'll have a look, thanks!

aspiers replied:
Too late - some of the necessary ingredients already got deleted: https://github.com/crowbar/crowbar-core/pull/1026#issuecomment-275105884

samirjorina commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="#C_EMAIL"
```
Should be "$C_EMAIL" instead of "#C_EMAIL"

jgrassler replied:
Fixed, thank you!

samirjorina commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="#C_EMAIL"
+
+alias Cpatch="(cd ~/crowbar-openstack; git show | patch -p1 --merge -d /opt/dell)"
```
I'd add -N to the patch command, so that it won't fail if patch was already applied

jgrassler replied:
Good idea, thanks. Added.

samirjorina commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+  barclamp_install.rb /opt/dell/crowbar_framework/barclamps/ &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ mybarclamp
```
use $barclamp variable instead of hard-coded mybarclamp

jgrassler replied:
Good point. Fixed. Also, I did the same for `/root/crowbar-openstack` while I was at it.

matrixik commented:
I think it would also be valuable to have short section about where to place rpm for development.

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+#### Adding Test Automation
+
+To include your barclamp in `mkcloud` test automation you would fork the
+[automation]() repository, add your barclamp to the list of barclamps enabled
```
Missing link.

jgrassler replied:
Fixed.

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+  snapshot of your admin node using its `createadminsnapshot` and
+  `restoreadminfromsnapshot` steps, respectively.
+
+* Create an additional OpenStack controller or compute node as a scratch node:
```
Best way for creating this additional node?
Could lonely node be used for this?
If https://github.com/SUSE-Cloud/automation/pull/1549 will be accepted and merged then lonely node for development could be created with:
```
sudo nodenumberlonelynode=1 /path/to/mkcloud setuplonelynodes crowbar_register
```

jgrassler replied:
Looks good. Thanks :-)

I added this to the documentation in slightly modified form.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+  components to this node while all other OpenStack services reside on the main
+  controller. This greatly shortens the time a chef run for your barclamp's
+  controller side components takes and thus speeds up development. Make a
+  backup of this node as well to guard against breakage and allow for
```
I'll add this as soon as there actually is generic backup-just-the-node-I-specified  and restore-just-the-node-I-specified functionality in `mkcloud`. I plan on adding this, but so far it does not exist, yet :-)

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+  # means higher priority)
+  chef_order: 98
+
+  # FIXME: The following appear to be unused. They are neither loaded in
```
For now, yes :-)

More to the point, I want to leave it in there until either somebody who either knows more about Crowbar than me confirms these keys can indeed be omitted or I have tried it. Until then I prefer to leave the FIXME in place so I know this still needs addressing.

aspiers replied:
I'm 85% sure that this comment is misleading and that at least one if not two or all three of these are in use.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+  `/opt/dell` (use this after each new commit). This command may cause merge
+  conflicts. If this happens, resolve them before doing anything else.
+
+* `sync-crowbar` will make any updated state in `/opt/dell` appear in the
```
Thanks, fixed!

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+* "Any Ceph OSD node"
+
+We have a limited range of node types in Crowbar to define types of roles.
+Beyond that
```
Ah, thanks. I left that for "later" because I first needed to look up a bunch of URLs. Guess I forgot about it later...fixed.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+        "api" : {
+           "bind_host" : "*",
+           "bind_port" : 12345,
+           "logfile" : "/var/log/myservice/myservice-api.log",
```
Done.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+
+When you are done `deployment[""]` should look something like this
+(this example assumes you have defined a `mybarclamp-compute` and a
+`-controller` role):
```
Done.

jgrassler commented on doc/barclamp.md:
````diff
@@ -0,0 +1,1177 @@
+@timeout = 3600
+
+main
+```
````
Done.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+First of all we create the directories
+
+* `crowbar_framework/app/views/barclamp/mybarclamp`
+* and `crowbar_framework/config/locales/mybarclamp`.
```
Agreed. That just sets people up for an `EPASTE`. Removed.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1177 @@
+  simple search-and-replace operation on another barclamp's controller)
+* `crowbar_framework/app/views/barclamp/mybarclamp/_edit_attributes.html.haml`
+* Errors in your default data bag, e.g.`chef/data_bags/crowbar/template-mybarclamp.json`
+  schema`. This is what error messages about invalid JSON syntax usually refer
```
Fixed.

matrixik commented on doc/barclamp.md:
```diff
@@ -5,6 +5,13 @@ through all the steps neccessary to create a minimal (without UI elements)
+## What is a Barclamp?
+
+A barclamp is a plugin for Crowbar that configures a particular service or
+group of services. It contains a che cookbook with one or more roles to
```
a Chef cookbook

jgrassler replied:
Thanks, fixed.

aspiers commented:
Thanks for this!  When it is merged, please can you add a link from https://github.com/SUSE/cloud/tree/master/docs#crowbar ?

samirjorina commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+
+sync_crowbar()
+  {
+  barclamp_install.rb $BARCLAMP_FORK_PATH &&
```
Here I'm using
`barclamp_install.rb /opt/dell/crowbar_framework/barclamps &&`

Otherwise I ended up with completely broken crowbar (problems with cinder, keystone, some weird migrations...). The suggested change simply 'reinstalls' existing barclamps and does not install them from git fork.

samirjorina commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+  barclamp_install.rb $BARCLAMP_FORK_PATH &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ $BARCLAMP_NAME
```
I guess this should be:
`knife cookbook upload -o /opt/dell/chef/cookbooks/ $barclamp`

jgrassler replied:
@aspiers: sure, will do!

@samirjorina: yes, there's a bit of a problem here indeed. Two problems actually:

The line you commented on causes the cookbook upload to fail completely. I don't think that would have caused the breakage you are observing, though.

That one is likely due to the second problem: I tried [Tom's suggestion](https://github.com/crowbar/crowbar/pull/2294#discussion_r93575431) a while ago and forgot to revert the `$BARCLAMP_FORK_PATH` in line 143  back to `/opt/dell/crowbar_framework/barclamps` before the next commit, effectively leaving the approach I tried there in place. Hence your problems with Cinder, Keystone and weird migrations are likely to be the situation described here: https://github.com/crowbar/crowbar/pull/2294#discussion_r94596298

Thank you for trying this and spotting the problem! I fixed both now.

aplanas commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by your first commit's ID.
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
```
What is this for?

jgrassler replied:
Yes, it's used to create the diff. One could use something like `git log master..HEAD --oneline | tail -n 1 | cut -f 1`, I suppose, but that relies on the state of the `master` branch being somewhat sensible: if somebody did their changes in their fork's `master` branch or merged part of their topic branch's history to their fork's `master` branch it will blow up. People being people this is likely to happen at some point, and a hardwired commit ID will still Just Work if somebody messes up their local `master` branch.

aspiers replied:
That doesn't work; maybe you mean `git log --format=%h -n1 master..HEAD`?  And you might need `--reverse` too; I'm not sure whether you want the oldest or newest commit.

jgrassler replied:
> That doesn't work

It does. It yields the oldest commit's short ID, which is what I use for creating the diff further down. Adding your `--format=%H` option would improve it a little in case of a duplicated short ID somewhere, but that's the extent to which it could be said not to work.

But the point is moot anyway: I don't want to rely on the `master` branch being in a usable state or somebody forgetting to switch the reference branch to the appropriate stable branch when doing a backport. It's a little bit of effort to dig up and paste the commit ID manually, but if it does go wrong it's easier to figure out where it went off the rails.

> I'm not sure whether you want the oldest or newest commit.

Good point. "first commit" is a rather ambiguous formulation. Fixed :-)

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+  forked [crowbar-openstack](https://www.github.com/crowbar/crowbar-openstack)
+  throughout the rest of this document.
+
+* A Crowbar admin node with the Crowbar version you are developing against
```
Would be good to mention here to install `git-core` and `patch` tools. New cloud created with `mkcloud` don't have them, maybe it `mkcloud` could install them automatically?

jgrassler replied:
Forgot to respond to this comment yesterday: good point, thanks. I always do this on autopilot so I forgot mentioning it. I added these two to the prerequisites list now.

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+alias Cpatch="(cd $BARCLAMP_FORK_PATH; git show | patch -N -p1 --merge -d /opt/dell)"
+
+# Full patch: applies your fork completely against /opt/crowbar
+function Fpatch()
```
Anyone else have problem with first run of `Fpatch` on new cloud creating file `/opt/dell/crowbar_framework/barclamps/*.yml` (yes, with `*` in the name)?
`sync_crowbar` fail with this file.

jgrassler replied:
I've had that happen before, but only once. I may have fixed it locally and forgotten to commit it. I'll try to reproduce it on a clean Crowbar node and fix the problem...

jgrassler replied:
I couldn't reproduce it this time around. I got rid of the wildcards, though, so this should no longer occur.

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+function Fpatch()
+  {
+  cd $BARCLAMP_FORK_PATH
+  git diff "$COMMIT^" HEAD | patch -N --merge -p1 -d /opt/dell
```
When I create new files this is working for me only the first time. Will create files in /opt/dell.
But running it second time it's showing:
```
The next patch would create the file chef/cookbooks/monasca/README.md,
which already exists!  Skipping patch.
1 out of 1 hunk ignored
```
for all new files. So new files are not updated with new content.
It will patch only already existing files:
```
patching file chef/cookbooks/barbican/recipes/api.rb
patching file chef/cookbooks/horizon/recipes/server.rb
```

jgrassler replied:
It's only supposed to be run once, with Cpatch being run after every new commit later. That _is_ a bit clumsy, though. I think I'll put some more effort into `Fpatch` to so it can do `Cpatch`'s job as well and not make a mess. Stay tuned :-)

jgrassler replied:
Alright, I replaced `Fpatch` by its idempotent cousin `reposync_crowbar` now (and got rid of `Cpatch` because `reposync_crowbar` will work in that situation too). Worked well for me, but feel free to give it a spin yourself.

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
+
+alias Cpatch="(cd $BARCLAMP_FORK_PATH; git show | patch -N -p1 --merge -d /opt/dell)"
```
Patching sometimes add same content many times to files in `/opt/dell` folder.
Sometimes files in `/opt/dell` end with git merge conflict markers.

jgrassler replied:
Yes. That would happen with multiple `Fpatch` invocations. See above. I'll try to figure out a way to simplify the whole patching bit.

jgrassler replied:
That should be fixed now with `reposync_crowbar` and its patch level record.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1224 @@
+fork of `crowbar-openstack` is out of sync with upstream.
+
+In the former case copying files from your fork to `/opt/dell` or using
+`barclapm_install` may cause an inconsistent state and errors entirely
```
Fixed.

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
```
Again, this would not be necessary if developing from your native dev environment, but if you are going to do this, it's better to do it via `git config`, so that it applies for all git commands, not just when you are running this script or running git commands in a shell where you forgot to first run these `export` commands.

jgrassler replied:
That is why I put these commands into .bashrc. I wanted to keep them in the environment because then they can be changed in a collaboration scenario without clobbering the hardwired configuration as set by `git config`.

jgrassler replied:
They have an identical copy of the script, since it's in the .bashrc. B can change the environment variables ad-hoc in just one terminal if the name/email address for A are named in the .bashrc. This ad-hoc approach won't change the persistent configuration, which prevents A from being surprised at suddenly committing as B.

I know that the same could be achieved by using `git config --global` and using environment variables only for an override. I opted for using the environment variable approach for the persistent configuration as well since it's a tiny little bit faster, not involving a file on disk and, more importantly, documents how to do the ad-hoc change for user B in a copy & paste friendly manner.

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+      return
+    fi
+  else
+    commit="${COMMIT}^"
```
IIUC, this will just end up pointing to `$branch` from above, so I'm not sure why you don't just ask the user to provide that value instead of manually calculating another commit from which you end up with the original answer.  Maybe I'm misunderstanding something - let's mumble quickly.

jgrassler replied:
Yes, of sorts. But given the uncertainties involved with diffing `$branch` against a stale `master` or stale `stable/3.0` I went for calculating the diff this way. This simply formats a range of commits as a patch, regardless of the (possibly stale) state of other branches.

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+
+  pushd $BARCLAMP_FORK_PATH > /dev/null
+  cp -a /opt/dell /opt/dell.orig
+  git diff $commit HEAD | patch -N --merge -p1 -d /opt/dell || return
```
This feels like a partial reinvention of [this wheel](https://github.com/crowbar/crowbar/blob/master/Guardfile).  I think I understand your reasons for using `patch`, but I disagree with them as explained below. Maybe I'm missing something though.

jgrassler replied:
You may be. Let me try and clarify my intentions behind re-inventing that particular wheel below. Maybe we'll find some common ground that way.

aspiers replied:
OK, sounds good!

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+
+In the former case copying files from your fork to `/opt/dell` or using
+`barclamp_install` may cause an inconsistent state and errors entirely
+unrelated to your change.
```
If that's the case then I'd argue that the development environment is fundamentally broken, since you're developing and testing an environment which will never appear again, as soon as your changes hit CI.  To be clear: you'd be developing and testing patches which are applied against an old commit, and then submitting them upstream as applied to the latest trunk.  A lot of the time you might get away with this without any major problems, but it still leaves you wide open to introducing bugs which were not caught during development and manual testing of the changes which are often more thorough regarding that particular code area than the CI will be (unless of course the code in question has good test coverage, but even then this makes me uncomfortable).

jgrassler replied:
Crowbar changes will take a while to percolate through CI and end up in the RPMs a development cloud is built from. Also, gating breakage happens and may hinder that percolation, sometimes for days on end. That way even a freshly built cloud can be out of date. Of course one should still give a patch a manual test run against a freshly built cloud, but the RPMs you get are occasionally behind Crowbar's `master` branch, and for these situations it is nice to have a tool that enables you to still continue development.

aspiers replied:
That kind of misses the point.  You're right that the *other* RPMs can lag behind `master`, but I'm talking about the RPM which you're actually hacking on in this context, which is presumably `crowbar-openstack`.  My point was that you're generating a patch from the topic branch which is based off the (merge-)base of the topic branch, and then applying that patch against whatever revision of `crowbar-openstack` happens to be deployed in `/opt/dell`.  That's akin to a three-way merge, and is hence prone to not only merge conflicts (which are OK, because they're visible breakage which can be fixed) but also invisible breakage due to incompatibilities between the two base versions in the merge.

Having said all that, Dirk reminded me earlier today that in this particular context of creating a new barclamp, it is far less likely that this kind of breakage would occur, because all we're doing really is adding new files.  So it would only break if there was some incompatible API change in the barclamp framework, and that's fairly unlikely most of the time.

So in summary, I'm not too worried about the downsides of your approach *in this context*, although I'd recommend thinking carefully before using that in other "normal" hacking contexts.

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+old code that is no longer present upstream. Both should not be an issue for
+new barclamps, but it is better to get used to the `patch(1)` based workflow
+right away since these problems do become a problem when changing an existing
+barclamp later and thus the `patch(1)` based workflow becomes neccessary.
```
Again, IMHO this is a broken workflow.  You should first rebase your topic branch so that it matches the `/opt/dell` which you are testing on.

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+
+To include your barclamp in `mkcloud` test automation you would fork the
+[automation](https://github.com/SUSE-Cloud/automation) repository, add your
+barclamp to the list of barclamps enabled by `qa_crobarsetup.sh` and submit a
```
`crobar` typo

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+place) you will have to add additional setup code to 
+[mkcloud](https://github.com/SUSE-Cloud/automation/blob/master/scripts/mkcloud)
+and/or
+[qa_crowbarsetup.sh](https://github.com/SUSE-Cloud/automation/blob/master/scripts/qa_crobarsetup.sh)
```
`crobar` typo

jgrassler replied:
Thanks, fixed!

aspiers commented on doc/barclamp.md:
````diff
@@ -0,0 +1,1321 @@
+
+```
+
+We will make heavy use of the `reposync_crowbar` command thus defined in the
````
Please can you take a look at https://github.com/crowbar/crowbar/blob/master/Guardfile and http://publishwith.me/crowbar-tree-merging and then consider whether a more unified approach might be possible?  (I'm not claiming it would be easy, but at least it would tackle the fundamental problem that the git repo structure doesn't match the deployed structure.)  If you'll forgive the history lesson, a long time ago in a universe far away, someone started to write a shell script a bit like this, with the goal of making Crowbar development smoother.  It grew into an 8,000-line monstrosity which could have been avoided by asking some more fundamental questions about development environments and workflows.  I'm not saying there's a similar risk here, but I think the KISS principle still applies :)  Would be very interested in your thoughts.

jgrassler replied:
> Please can you take a look at https://github.com/crowbar/crowbar/blob/master/Guardfile and
http://publishwith.me/crowbar-tree-merging and then consider whether a more unified approach might be possible? (I'm not claiming it would be easy, but at least it would tackle the fundamental problem that the git repo structure doesn't match the deployed structure.)

Not everybody will happy to perform
https://github.com/crowbar/crowbar/blob/master/Guardfile#L20 on their local
machine. But, more importantly, this approach is awkward in a collaboration scenario. That being said I added a link for those who prefer
working locally.

As for http://publishwith.me/crowbar-tree-merging: I'm cautiously optimistic
that we might have a unified solution at some stage, but the reality is that
we do not have a unified solution right now. Nor do I intend to provide one
with that little hack. It's just something a new developer can use for their
first steps (after that they'll build something of their own anyway, and _that_
may end up they unified solution we are looking for).

> If you'll forgive the history lesson, a long time ago in a universe far away, someone started to write a shell script a bit like this, with the goal of making Crowbar development smoother. It grew into an 8,000-line monstrosity which could have been avoided by asking some more fundamental questions about development environments and workflows. I'm not saying there's a similar risk here, but I think the KISS principle still applies :) Would be very interested in your thoughts.

I bore the KISS principle in mind as I came up with this :-)

While the script grew some tentacles, still I tried to limit these as far as
possible. Also, I put it in here on purpose: I consider it more
documentation/inspiration than code to be run as-is since it shows what needs
to be done to synchronize the state of one's `crowbar-openstack` fork with a
running Crowbar instance. I only intend to change it if these synchronization
needs change. It is definitely not supposed to solve everybody's development
environment problems. Where that lofty goal is concerned its only aspiration is
showing people what challenges they face if they chose to work on that solution.

jgrassler replied:
"lowest common denominator" captures it pretty well, yes :-)

For any 10 developers you'll have about 25 vastly differing local development environments. Instead of trying to come up with something all of them are happy with I wanted to accomplish two things:

* Provide something anybody can use even if their workstation accidently got an anvil dropped on it an hour ago and now they have a new machine fresh from IT with none of their customized personal environment on it.
* Provide something that shows them all the bits they need to touch in a compact, easy to understand manner so they can roll their own development environment if they need to or fix the Guardfile themselves if it bitrots.

aspiers commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+it doesn't appear or you got an error message from the `sync_crowbar` command
+you will now have to start debugging. Common errors at this stage are mostly
+bad file names, or problems in `mybarclamp.yml`.
+
```
I think there's a missing step here, or somewhere around here.  You cover local testing via some mechanism for quickly transferring changes from a git repo into `/opt/dell` (and you suggest `reposync_crowbar` which is one possibility but I have also mentioned others).  This speed of transferral is of course important to keep the hack/test cycle efficient.  So far so good.  But when the changes are stabilising, it then makes sense to sacrifice some development speed for higher quality testing, and in general it's more reliable to test by deploying rpms, which get the test deployment closer to a real world deployment.

The downside of testing via rpm deployments is that it takes a lot longer to build an rpm from a git topic branch and then deploy it to a local test cloud.  But that's exactly why [I automated this process down to a one-line command!](https://github.com/openSUSE/pack-tools/tree/master/contrib/BS-pkg-testing)  So I would like to see that also covered in this guide.

jgrassler replied:
Oh. I didn't know that one :-)

Thanks!

I included a link to that as well.

aspiers replied:
Cool, thanks :)

aspiers commented:
BTW, once this gets merged, please ensure it gets linked from https://github.com/SUSE/cloud/tree/master/docs - thanks! /cc @nkrinner

nkrinner commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+  `template-mybarclamp.schema`. This is the most common error.
+
+* Logical errors in `crowbar_framework/app/models/mybarclamp_service.rb`. These
+  may most commonly are role constraint mismatches or fields required by the
```
Not a native speaker, but the 'may' can be dropped?

aspiers replied:
Yep.  "These are most commonly"

matrixik commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+new barclamps, but it is better to get used to the `patch(1)` based workflow
+right away since these problems do become a problem when changing an existing
+barclamp later and thus the `patch(1)` based workflow becomes neccessary.
+
```
Then maybe you could share what is your workflow?

dirkmueller commented:
@jgrassler have a chance to address the comments? maybe we could merge it then.

jgrassler commented:
I'm dashing back and forth between a bunch of right now, so probably not today. I'll just make the time to address them tomorrow, though. I want to get this moving along, too...

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
```
Yes. In fact this is what I am counting on. It will provide a uniform environment where editor, configuration files, ssh keys (or the lack thereof are known quantities).

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
```
It would be nicer and developers are free to work that way. As a documentation writer that presents me with a moving target, though.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
```
Yes. This makes collaboration that much harder, though. One of the main aims of this approach is providing an environment everyone can live with and use in a collaboration scenario on a remote Crowbar node.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+# to the user to set $COMMIT as they see fit rather than trying to determine
+# its value automatically through the command mentioned above.
+
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
```
There may be multiple commits in a topic branch. Not everyone uses the OpenStack model of `git commit --amend` that leaves just a single commit in a topic branch. It doesn't make sense without Gerrit or similar heavy CI artillery. Likewise, `master` or `stable/3.0` may be way out of date in a fork (it commonly is). The raw SHA1 guarantees you a diff from your first commit to HEAD on the topic branch. Nothing more, nothing less. That is why I opted for this.

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+manual installation or by using an automated setup tool such as
+[mkcloud](https://github.com/SUSE-Cloud/automation/blob/master/scripts/mkcloud).
+Depending on the individual situation you can use either the `plain` or the
+`installonly` `mkcloud` step. The former will install the nodes and deploy the
```
Thanks, fixed!

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+standard set of barclamps in their default configuration, while the latter will
+only install the nodes. Use `plain` if your barclamp does not need to be
+deployed before other barclamps and if you do not modify other barclamps.
+Otherwise use `installonly` and deploy barclamps bit by bit through the Crowbar
```
Thanks, fixed!

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+  actually deploying your service.
+* Some glue code to tie your chef cookbook into the Crowbar Web UI. This glue
+  code will parametrize your chef cookbook and apply its component roles to the
+  nodes it orchestrates.
```
Ok, fixed

jgrassler commented on doc/barclamp.md:
```diff
@@ -0,0 +1,1321 @@
+        },
+         "mybarclamp-compute" => {
+          "unique" => false,
+          "count" => -1,   # Can be deployed to an arbitrary number of Compute nodes
```
Yes...I wanted to have a concrete usage example as well, in case somebody missed the comment above.

jgrassler commented:
Alright, I think I addressed/discussed all current comments now. Let's see if we can reach a consensus on this...

Also, thank you for your suggestion on the [Lx:Ly style links](https://github.com/jgrassler/crowbar/pull/1), @sjamgade! I integrated it, but I had already rebased when I thought of the still open pull request.

aspiers commented:
Just discussed this with @jgrassler and we agreed he will:

- move the info about the option of using a local develop environment so that it appears before the detailed instructions on setting up a dev environment on the admin server
- drop all the `$COMMIT` stuff and instead use the 3-dot form of `git diff` to automatically generate the patch from the merge-base of the topic branch
- fix the issue with the `cp` to a `.bak` directory

After that, I'll be an enthusiastic +1 :-)

aspiers commented on doc/barclamp.md:
```diff
@@ -85,24 +124,14 @@ nodenumberlonelynode=1 mkcloud setuplonelynodes crowbar_register
-Crowbar node and switch to your development branch.
+Crowbar node and switch to your development branch. Note, that the code below
+assumes you are working on a dedicated topic branch. It will break if you are
+doing development on your fork's `master` branch.
```
Yes, you could even be more forceful and say "DON'T WORK ON master!" :)

jgrassler replied:
Done.

aspiers commented on doc/barclamp.md:
```diff
@@ -168,14 +181,35 @@ function reposync_crowbar()
+    echo " and remove /opt/dell/back.bak once you have done so." 1>&2
+    return 1
+  fi
+  if [ -n "$commit" ]; then  # only apply commits added since the last reposync_crowbar run
```
Might be worth a comment that this can only get set via `/opt/dell/patchlevel` now.

jgrassler replied:
Done.
